### PR TITLE
Add support for subject alternative names when requesting certificates

### PR DESCRIPTION
**Pull request by ean**

### CODE REVIEW

**Comment by luna-duclos** [comment]
This will need to wait until xenolf/lego#356 is merged

**Comment by whereisaaron** [comment]
Looks great @ean 👏 

Thanks @luna-duclos I was going to note the same thing. I saw ACMESharp users were also having trouble with SAN certificates and this rate limit. Hopefully xenolf/lego#356 will fix that.

> The “new-reg”, “new-authz” and “new-cert” endpoints have an Overall Requests limit of 20 per second. All other endpoints have an Overall Requests limit of 2000 per second.

There is this newly introduced failures rate limit to watch out for, if we retry failed validations too quickly. But that's nothing specific to this PR.

> We will soon (February 2017) introduce a Failed Validation limit of 5 failures per account, per hostname, per hour. This limit will be higher on staging so you can use staging to debug connectivity problems. 

https://letsencrypt.org/docs/rate-limits/

**Comment by luna-duclos** [comment]
xenolf/lego's master should now have a fix for the rate limit.

**Comment by luna-duclos** [comment]
@ean: Are there any further changes you would like to make to this before I review/merge it ?

**Comment by ean** [comment]
@luna-duclos please review, I have been running this at work for about a month now, and it seems stable.

**Comment by rosskukulinski** [comment]
I've been running @ean's code for SANs as well as a patched xenolf/lego for a few weeks now and haven't had any problems.  Thank you @ean!

File: processor.go
Comment (euank) [inline processor.go]: This for loop could be `copy(domains[1:], altNames)` instead; slightly more readable I think.

Comment (euank) [reply processor.go]: Actually, after thinking a second more, I think the most readable option is:

`domains := append([]string{cert.Spec.Domain}, altNames...)`

File: processor.go
Comment (euank) [inline processor.go]: It's probably okay to just slice `tls.Hosts`; `altNames = tls.Hosts[1:]` The tls object should be considered immutable anyways (and indeed processCertificate does copy it).

File: processor.go
Comment (euank) [inline processor.go]: it seems like it would be easier to `return []string{}, nil` here rather than go to the trouble of marshal+unmarshal to do the same.

**Review by euank — COMMENTED** [review]
I added a few style nits, but this overall looks good I think.

The one other thing to consider would be updating the glide.lock file (and possibly checking in the vendor tree) since apparently there are upstream lego fixes that would help.

**Comment by ean** [comment]
Thanks @euank, updated the code to match your comments. It ended up being a lot easier on the eyes.

**Comment by euank** [comment]
This PR also needs to remove the length check here: https://github.com/PalmStoneGames/kube-cert-manager/blob/bc5064a21b1c9ec0c21590ce41b47526884ca647/processor.go#L619-L621

Without that change, I have it erroneously delete the secret for an ingress the next sync-interval after creating it.

Maybe it should just be flipped to a `< 1` check to ensure the bit below doesn't panic.

**Comment by ean** [comment]
@euank - Thanks for testing this. I updated the ingress certificate logic now. I also saw that AltNames were not added to the certificate spec, which I also fixed. Could you give it another test? I don't have any ingress setup to test this on.

**Comment by euank** [comment]
I'm not seeing any other issues, though I'm running it rebased on top of my client-go change as well.

**Comment by euank** [comment]
Ping @luna-duclos, what's the next step for this?

**Comment by luna-duclos** [comment]
I've been away for easter! I will be merging those, and the other open PRs, pretty soon, then we can have a test period, and then when they prove stable, make a new release and move the repo.

File: processor.go
Comment (luna-duclos) [inline processor.go]: Can just return nil here instead of an empty initialized slice

File: processor.go
Comment (luna-duclos) [inline processor.go]: This error should be wrapped

**Review by luna-duclos — APPROVED** [review]
Two small comments, if you don't mind looking at them, I don't mind fixing them myself either if you lack the time, let me know! :)

File: processor.go
Comment (ean) [reply processor.go]: fixed

File: processor.go
Comment (ean) [reply processor.go]: fixed

**Pull request merged**
